File: src/userservice/userservice/controller.py
from grpc import StatusCode

from userservice import thoughts_pb2_grpc, thoughts_pb2
from userservice.crypto import generate_hash, validate_password
from userservice.utils import is_valid_email
from userservice import logger


class Controller(thoughts_pb2_grpc.UserServiceServicer):
    def __init__(self, db_client):
        self.db_client = db_client

    def CreateUser(self, request, context):
        if (len(request.username) == 0
            or len(request.email) == 0
            or len(request.password) == 0):
            context.abort(
                StatusCode.INVALID_ARGUMENT,
                'Name, username, email and password are required.'
            )
        elif is_valid_email(request.email) == False:
            context.abort(
                StatusCode.INVALID_ARGUMENT,
                'Invalid email address.'
            )

        try:
            result = self.db_client.create_user(
                request.name, request.username,
                request.email, generate_hash(request.password))
            return thoughts_pb2.Identifier(id=result)
        except Exception as e:
            logger.print(f'Creating user failed: {e}')
            context.abort(
                StatusCode.INVALID_ARGUMENT,
                'User with this username or email already exists.'
            )

    def GetUser(self, request, context):
        user_id = dict(context.invocation_metadata())['user-id']

        try:
            result = self.db_client.get_user(
                request.user_id, request.username, user_id)
        except Exception as e:
            logger.print(f'Getting user failed: {e}')
            context.abort(StatusCode.INTERNAL)

        if result is None:
            context.abort(StatusCode.NOT_FOUND)
        return result

    def UpdateUser(self, request, context):
        user_id = dict(context.invocation_metadata())['user-id']

        if len(request.password) != 0:
            return self.updatePassword(request, context)

        if is_valid_email(request.email) == False:
            context.abort(
                StatusCode.INVALID_ARGUMENT,
                'Invalid email address.'
            )

        try:
            self.db_client.update_user(
                user_id, request.name, request.username,
                request.email, request.bio)
            return thoughts_pb2.Empty()
        except Exception as e:
            logger.print(f'Updating user failed: {e}')
            context.abort(StatusCode.INTERNAL)

    def updatePassword(self, request, context):
        user_id = dict(context.invocation_metadata())['user-id']

        if len(request.old_password) == 0:
            context.abort(
                StatusCode.INVALID_ARGUMENT,
                'Both password and the current password are required.'
            )

        try:
            result = self.db_client.get_user_with_id(user_id)
        except Exception as e:
            logger.print(f'Getting user failed: {e}')
            context.abort(StatusCode.INTERNAL)

        if validate_password(request.password, result['password']) == False:
            context.abort(
                StatusCode.INVALID_ARGUMENT,
                'Wrong password. Enter the correct current password.'
            )

        try:
            self.db_client.update_password(
                user_id, generate_hash(request.password))
            return thoughts_pb2.Empty()
        except Exception as e:
            logger.print(f'Updating user failed: {e}')
            context.abort(StatusCode.INTERNAL)

    def GetFollowing(self, request, context):
        user_id = dict(context.invocation_metadata())['user-id']

        try:
            result = self.db_client.get_following(
                request.user_id, request.page,
                request.limit, user_id)
            return thoughts_pb2.Users(users=result)
        except Exception as e:
            logger.print(f'Getting users failed: {e}')
            context.abort(StatusCode.INTERNAL)

    def GetFollowers(self, request, context):
        user_id = dict(context.invocation_metadata())['user-id']

        try:
            result = self.db_client.get_followers(
                request.user_id, request.page,
                request.limit, user_id)
            return thoughts_pb2.Users(users=result)
        except Exception as e:
            logger.print(f'Getting users failed: {e}')
            context.abort(StatusCode.INTERNAL)

    def FollowUser(self, request, context):
        user_id = dict(context.invocation_metadata())['user-id']

        try:
            self.db_client.follow_user(request.user_id, user_id)
            return thoughts_pb2.Empty()
        except Exception as e:
            logger.print(f'Following user failed: {e}')
            context.abort(StatusCode.INTERNAL)

    def UnfollowUser(self, request, context):
        user_id = dict(context.invocation_metadata())['user-id']

        try:
            self.db_client.unfollow_user(request.user_id, user_id)
            return thoughts_pb2.Empty()
        except Exception as e:
            logger.print(f'Unfollowing user failed: {e}')
            context.abort(StatusCode.INTERNAL)

File: src/apigateway/service/user/controller.go
package user

import (
	"context"
	"log"
	"net/http"
	"strconv"
	"time"

	"google.golang.org/grpc"

	"github.com/labstack/echo/v4"
	pb "github.com/robbdimitrov/thoughts/src/apigateway/genproto"
	"github.com/robbdimitrov/thoughts/src/apigateway/service/utils"
)

type Controller struct {
	addr string
}

func NewController(addr string) *Controller {
	return &Controller{addr: addr}
}

func (s *Controller) createUser(w http.ResponseWriter, r *http.Request) {
	conn, err := grpc.Dial(s.addr, utils.InsecureCredentials(), grpc.WithBlock())
	if err != nil {
		log.Printf("Connecting to service failed: %v", err)
		// utils.ToHTTPError(http.StatusInternalServerError)
		return
	}
	defer conn.Close()
	client := pb.NewUserServiceClient(conn)

	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	defer cancel()

	req := pb.CreateUserRequest{
		// Name:     c.FormValue("name"),
		// Username: c.FormValue("username"),
		// Email:    c.FormValue("email"),
		// Password: c.FormValue("password"),
	}

	_, err = client.CreateUser(ctx, &req)
	if err != nil {
		log.Printf("Creating user failed: %v", err)
		utils.ToHTTPError(err)
		return
	}

	// return c.JSON(201, map[string]int32{"id": res.Id})
}

func (s *Controller) getUser(w http.ResponseWriter, r *http.Request) {
	conn, err := grpc.Dial(s.addr, utils.InsecureCredentials(), grpc.WithBlock())
	if err != nil {
		log.Printf("Connecting to service failed: %v", err)
		return echo.NewHTTPError(500)
	}
	defer conn.Close()
	client := pb.NewUserServiceClient(conn)

	userID, err := strconv.Atoi(c.Param("userId"))
	if err != nil {
		return echo.NewHTTPError(400)
	}

	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	ctx = appendUserIDHeader(ctx, c)
	defer cancel()

	req := pb.UserRequest{UserId: int32(userID)}

	res, err := client.GetUser(ctx, &req)
	if err != nil {
		log.Printf("Getting user failed: %v", err)
		return utils.ToHTTPError(err)
	}

	return c.JSON(200, mapUser(res))
}

func (s *Controller) updateUser(w http.ResponseWriter, r *http.Request) {
	conn, err := grpc.Dial(s.addr, utils.InsecureCredentials(), grpc.WithBlock())
	if err != nil {
		log.Printf("Connecting to service failed: %v", err)
		return echo.NewHTTPError(500)
	}
	defer conn.Close()
	client := pb.NewUserServiceClient(conn)

	currentUserID := getUserID(c)
	if currentUserID != c.Param("userId") {
		return echo.NewHTTPError(403)
	}

	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	ctx = appendUserIDHeader(ctx, c)
	defer cancel()

	req := pb.UpdateUserRequest{
		Name:        c.FormValue("name"),
		Username:    c.FormValue("username"),
		Email:       c.FormValue("email"),
		Password:    c.FormValue("password"),
		OldPassword: c.FormValue("oldPassword"),
	}

	_, err = client.UpdateUser(ctx, &req)
	if err != nil {
		log.Printf("Updating user failed: %v", err)
		return newHTTPError(err)
	}

	return c.NoContent(http.StatusNoContent)
}

func (s *Controller) getFollowing(w http.ResponseWriter, r *http.Request) {
	conn, err := grpc.Dial(s.addr, insecureCredentials(), grpc.WithBlock())
	if err != nil {
		log.Printf("Connecting to service failed: %v", err)
		return echo.NewHTTPError(500)
	}
	defer conn.Close()
	client := pb.NewUserServiceClient(conn)

	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	ctx = appendUserIDHeader(ctx, c)
	defer cancel()

	userID, err := strconv.Atoi(c.Param("userId"))
	if err != nil {
		return echo.NewHTTPError(400)
	}
	page, err := strconv.Atoi(c.QueryParam("page"))
	if err != nil {
		return echo.NewHTTPError(400)
	}
	limit, err := strconv.Atoi(c.QueryParam("limit"))
	if err != nil {
		return echo.NewHTTPError(400)
	}

	req := pb.GetUsersRequest{
		UserId: int32(userID),
		Page:   int32(page),
		Limit:  int32(limit),
	}

	res, err := client.GetFollowing(ctx, &req)
	if err != nil {
		log.Printf("Getting following failed: %v", err)
		return newHTTPError(err)
	}

	users := make([]user, len(res.Users))
	for i, v := range res.Users {
		users[i] = mapUser(v)
	}

	return c.JSON(200, map[string][]user{"items": users})
}

func (s *Controller) getFollowers(w http.ResponseWriter, r *http.Request) {
	conn, err := grpc.Dial(s.addr, insecureCredentials(), grpc.WithBlock())
	if err != nil {
		log.Printf("Connecting to service failed: %v", err)
		return echo.NewHTTPError(500)
	}
	defer conn.Close()
	client := pb.NewUserServiceClient(conn)

	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	ctx = appendUserIDHeader(ctx, c)
	defer cancel()

	userID, err := strconv.Atoi(c.Param("userId"))
	if err != nil {
		return echo.NewHTTPError(400)
	}
	page, err := strconv.Atoi(c.QueryParam("page"))
	if err != nil {
		return echo.NewHTTPError(400)
	}
	limit, err := strconv.Atoi(c.QueryParam("limit"))
	if err != nil {
		return echo.NewHTTPError(400)
	}

	req := pb.GetUsersRequest{
		UserId: int32(userID),
		Page:   int32(page),
		Limit:  int32(limit),
	}

	res, err := client.GetFollowers(ctx, &req)
	if err != nil {
		log.Printf("Getting followers failed: %v", err)
		return newHTTPError(err)
	}

	users := make([]user, len(res.Users))
	for i, v := range res.Users {
		users[i] = mapUser(v)
	}

	return c.JSON(200, map[string][]user{"items": users})
}

func (s *Controller) followUser(w http.ResponseWriter, r *http.Request) {
	conn, err := grpc.Dial(s.addr, utils.InsecureCredentials(), grpc.WithBlock())
	if err != nil {
		log.Printf("Connecting to service failed: %v", err)
		// echo.NewHTTPError(http.StatusInternalServerError)
		return
	}
	defer conn.Close()
	client := pb.NewUserServiceClient(conn)

	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	// ctx = appendUserIDHeader(ctx, c)
	defer cancel()

	userID, err := strconv.Atoi(c.Param("userId"))
	if err != nil {
		// echo.NewHTTPError(http.StatusBadRequest)
		return
	}
	req := pb.UserRequest{UserId: int32(userID)}

	_, err = client.FollowUser(ctx, &req)
	if err != nil {
		log.Printf("Following user failed: %v", err)
		// newHTTPError(err)
		return
	}

	return c.NoContent(http.StatusNoContent)
}

func (s *Controller) unfollowUser(w http.ResponseWriter, r *http.Request) {
	conn, err := grpc.Dial(s.addr, utils.InsecureCredentials(), grpc.WithBlock())
	if err != nil {
		log.Printf("Connecting to service failed: %v", err)
		// echo.NewHTTPError(http.StatusInternalServerError)
		return
	}
	defer conn.Close()
	client := pb.NewUserServiceClient(conn)

	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	// ctx = appendUserIDHeader(ctx, c)
	defer cancel()

	userID, err := strconv.Atoi(c.Param("userId"))
	if err != nil {
		echo.NewHTTPError(http.StatusBadRequest)
		return
	}
	req := pb.UserRequest{UserId: int32(userID)}

	_, err = client.UnfollowUser(ctx, &req)
	if err != nil {
		log.Printf("Unfollowing user failed: %v", err)
		utils.ToHTTPError(err)
		return
	}

	// return c.NoContent(http.StatusNoContent)
}
